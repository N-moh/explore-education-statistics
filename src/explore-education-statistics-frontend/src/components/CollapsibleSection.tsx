import classNames from 'classnames';
import React, { Component, ReactNode, SyntheticEvent } from 'react';
import styles from './CollapsibleSection.module.scss';

interface Props {
  children: ReactNode;
  heading: ReactNode | string;
  contentId?: string;
  open?: boolean;
}

interface State {
  isCollapsed: boolean;
}

let idCounter = 0;

class CollapsibleSection extends Component<Props, State> {
  public static defaultProps = {
    open: false,
  };

  public state = {
    contentId:
      this.props.contentId || `collapsible-section-${(idCounter += 1)}`,
    isCollapsed: !this.props.open,
  };

  public render() {
    const { heading, children } = this.props;
    const { isCollapsed, contentId } = this.state;

    return (
      <div>
        <h3 className={styles.heading}>
          <button
            aria-controls={contentId}
            aria-expanded={!isCollapsed}
            className={styles.collapseToggle}
            onClick={this.handleClick}
          >
            {heading}
          </button>
        </h3>

        <div
          className={classNames(styles.content, {
            [styles.collapsed]: isCollapsed,
          })}
          hidden={isCollapsed}
          id={contentId}
        >
          {children}
        </div>
      </div>
    );
  }

  private handleClick = (e: SyntheticEvent) => {
    e.preventDefault();

    this.setState({
      isCollapsed: !this.state.isCollapsed,
    });
  };
}

export default CollapsibleSection;
