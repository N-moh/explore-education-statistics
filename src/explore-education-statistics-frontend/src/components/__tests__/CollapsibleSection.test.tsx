import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import CollapsibleSection from '../CollapsibleSection';

describe('CollapsibleSection', () => {
  test('renders with hidden content by default', () => {
    const { container } = render(
      <CollapsibleSection heading="Test heading" contentId="the-content">
        <p>Test content</p>
      </CollapsibleSection>,
    );

    expect(container.querySelector('#the-content')).toHaveAttribute('hidden');
    expect(container).toMatchSnapshot();
  });

  test('renders with visible content when `open` is true', () => {
    const { container } = render(
      <CollapsibleSection heading="Test heading" contentId="the-content" open>
        <p>Test content</p>
      </CollapsibleSection>,
    );

    expect(container.querySelector('#the-content')).not.toHaveAttribute(
      'hidden',
    );
    expect(container).toMatchSnapshot();
  });

  test('clicking heading makes the hidden content visible', () => {
    const { container, getByText } = render(
      <CollapsibleSection heading="Test heading" contentId="the-content">
        <p>Test content</p>
      </CollapsibleSection>,
    );

    const content = container.querySelector('#the-content') as HTMLElement;

    expect(content).toHaveAttribute('hidden');

    fireEvent.click(getByText('Test heading'));

    expect(content).not.toHaveAttribute('hidden');
  });

  test('auto-generates content `id` if no `contentId` is set', () => {
    const { container } = render(
      <CollapsibleSection heading="Test heading">
        <p>Test content</p>
      </CollapsibleSection>,
    );

    const content = container.querySelector('[id]') as HTMLElement;

    expect(content.getAttribute('id')).toContain('collapsible-section-');
  });
});
