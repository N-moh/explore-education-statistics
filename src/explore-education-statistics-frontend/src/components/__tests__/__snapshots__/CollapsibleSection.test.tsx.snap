// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleSection renders with hidden content by default 1`] = `
<div>
  <div>
    <h3
      class="heading"
    >
      <button
        aria-controls="the-content"
        aria-expanded="false"
        class="collapseToggle"
      >
        Test heading
      </button>
    </h3>
    <div
      class="content collapsed"
      hidden=""
      id="the-content"
    >
      <p>
        Test content
      </p>
    </div>
  </div>
</div>
`;

exports[`CollapsibleSection renders with visible content when \`open\` is true 1`] = `
<div>
  <div>
    <h3
      class="heading"
    >
      <button
        aria-controls="the-content"
        aria-expanded="true"
        class="collapseToggle"
      >
        Test heading
      </button>
    </h3>
    <div
      class="content"
      id="the-content"
    >
      <p>
        Test content
      </p>
    </div>
  </div>
</div>
`;
